src/rgw/rgw_rest_swift.cc

```diff
@@ -1350,38 +1355,131 @@ RGWOp *RGWHandler_REST_Obj_SWIFT::op_options()
 
 int RGWHandler_REST_SWIFT::authorize()
 {
-  if ((!s->os_auth_token && s->info.args.get("temp_url_sig").empty()) ||
-      (s->op == OP_OPTIONS)) {
-    /* anonymous access */
-    rgw_get_anon_user(*(s->user));
-    s->perm_mask = RGW_PERM_FULL_CONTROL;
-    return 0;
-  }
+  /* Factories. */
+  class SwiftAuthFactory
+    : public RGWTempURLAuthApplier::Factory,
+      public RGWLocalAuthApplier::Factory,
+      public RGWRemoteAuthApplier::Factory {
+    typedef RGWAuthApplier::aplptr_t aplptr_t;
+
+    RGWRados * const store;
+    const std::string acct_override;
+
+  public:
+    SwiftAuthFactory(RGWRados * const store,
+                     const std::string& acct_override)
+      : store(store),
+        acct_override(acct_override) {
+    }
+
+    aplptr_t create_apl_turl(CephContext * const cct,
+                             const RGWUserInfo& user_info) const override {
+      /* TempURL doesn't need any user account override. It's a Swift-specific
+       * mechanism that requires  account name internally, so there is no
+       * business with delegating the responsibility outside. */
+      return aplptr_t(new RGWTempURLAuthApplier(cct, user_info));
+    }
 
-  bool authorized = rgw_swift->verify_swift_token(store, s);
-  if (!authorized)
-    return -EPERM;
+    aplptr_t create_apl_local(CephContext * const cct,
+                           const RGWUserInfo& user_info,
+                           const std::string& subuser) const override {
+      return aplptr_t(
+        new RGWThirdPartyAccountAuthApplier<RGWLocalAuthApplier>(
+          RGWLocalAuthApplier(cct, user_info, subuser),
+          store, acct_override));
+    }
 
-  if (s->user->system) {
-    s->system_request = true;
-    ldout(s->cct, 20) << "system request over Swift API" << dendl;
+    aplptr_t create_apl_remote(CephContext * const cct,
+                               RGWRemoteAuthApplier::acl_strategy_t&& acl_alg,
+                               const RGWRemoteAuthApplier::AuthInfo info
+                              ) const override {
+      return aplptr_t(
+        new RGWThirdPartyAccountAuthApplier<RGWRemoteAuthApplier>(
+          RGWRemoteAuthApplier(cct, store, std::move(acl_alg), info),
+          store, acct_override));
+    }
+  } aplfact(store, s->account_name);
+
+  /* Extractors. */
+  RGWXAuthTokenExtractor token_extr(s);
+
+  /* Auth engines. */
+  RGWTempURLAuthEngine tempurl(s, store,                    &aplfact);
+  RGWSignedTokenAuthEngine rgwtk(s->cct, store, token_extr, &aplfact);
+  RGWKeystoneAuthEngine keystone(s->cct,        token_extr, &aplfact);
+  RGWExternalTokenAuthEngine ext(s->cct, store, token_extr, &aplfact);
+  RGWAnonymousAuthEngine anoneng(s->cct,                    &aplfact);
+
+  /* Pipeline. */
+  const std::vector<const RGWAuthEngine *> engines = {
```

Comment:
I think `std::vector` might be replaced with `boost::static_vector` or similar. Is it available for us (taking into consideration boost version and platforms we do support)?